Hw10 program optimization is completed

novikov-ai opened the pull request:
## Домашнее задание №10 «Оптимизация программы»

> Benchstat comparison:
name              old time/op    new time/op    delta
GetDomainStat-12    7.68ms ±17%    8.28ms ± 3%     ~     (p=0.151 n=5+5)
name              old alloc/op   new alloc/op   delta
GetDomainStat-12    2.38MB ±18%    0.18MB ± 5%  -92.40%  (p=0.008 n=5+5)
name              old allocs/op  new allocs/op  delta
GetDomainStat-12     23.2k ±18%      5.6k ± 5%  -75.75%  (p=0.008 n=5+5)

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я удалил `.sync` файл. [Зачем его удалять?](https://github.com/OtusGolang/home_work/wiki/%5B%D0%A1%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0%D0%BC%5D-%D0%9F%D1%80%D0%BE%D1%86%D0%B5%D1%81%D1%81-%D1%81%D0%B4%D0%B0%D1%87%D0%B8-%D0%94%D0%97#user-content-%D0%92%D0%B0%D1%80%D0%B8%D0%B0%D0%BD%D1%82-2)
- [x] Я написал бенчмарк и приложил результаты benchstat.

### Критерии оценки
- [x] Пайплайн зелёный и нет попытки «обмануть» систему - 4 балла
- [x] Добавлены юнит-тесты - до 3 баллов
- [x] Понятность и чистота кода - до 3 баллов

#### Зачёт от 7 баллов

novikov-ai committed:
init

novikov-ai committed:
improve performance

novikov-ai committed:
clean up

novikov-ai committed:
lint

novikov-ai committed:
lint

agneum commented on hw10_program_optimization/stats.go:
```diff
@@ -32,35 +43,53 @@ func GetDomainStat(r io.Reader, domain string) (DomainStat, error) {
+
+		user := getUserFromParsedValue(v)
+
+		result[i] = *user
```
В рамках задачи можно не сохранять данные пользователей и сразу делать проверку и инкремент счётчика домена

agneum commented on hw10_program_optimization/stats.go:
```diff
@@ -32,35 +43,53 @@ func GetDomainStat(r io.Reader, domain string) (DomainStat, error) {
-		if err != nil {
-			return nil, err
-		}
+	regExpr, err := regexp.Compile("\\." + domain)
```
Предпочтительнее использовать `strings.HasSuffix` вместо регулярного выражения

agneum commented on hw10_program_optimization/stats_test.go:
```diff
@@ -36,4 +38,40 @@ func TestGetDomainStat(t *testing.T) {
+}
+
+func BenchmarkGetDomainStat(b *testing.B) {
+	r, _ := zip.OpenReader("testdata/users.dat.zip")
```
Ошибки важно обрабатывать

agneum commented on hw10_program_optimization/stats_test.go:
```diff
@@ -36,4 +38,40 @@ func TestGetDomainStat(t *testing.T) {
+	r, _ := zip.OpenReader("testdata/users.dat.zip")
+	defer r.Close()
+
+	data, _ := r.File[0].Open()
```
Рекомендую использовать https://pkg.go.dev/testing#B.StopTimer и https://pkg.go.dev/testing#B.StartTimer, чтобы исключить время на подготовку данных

agneum reviewed (approved):
Спасибо за выполненное задание. Хорошая работа
Итого: 10 баллов из 10

novikov-ai committed:
add error handle & start/stop benchTimer for preparing data

novikov-ai committed:
improve performance:
name              old time/op    new time/op    delta
GetDomainStat-12    8.28ms ± 3%    9.18ms ± 0%  +10.93%  (p=0.016 n=5+4)
name              old alloc/op   new alloc/op   delta
GetDomainStat-12     181kB ± 5%      31kB ±12%  -82.76%  (p=0.008 n=5+5)
name              old allocs/op  new allocs/op  delta
GetDomainStat-12     5.62k ± 5%     1.02k ±14%  -81.89%  (p=0.008 n=5+5)

novikov-ai committed:
Merge branch 'master' into hw10_program_optimization
